components/suggestion/suggestion_box/suggestion_box.tsx

```diff
@@ -0,0 +1,776 @@
+    terms: ProviderSuggestions['terms'];
+    components: Array<React.Component<{item: Record<string, unknown>}>>;
+    selection: string;
+    selectionIndex: number;
+    allowDividers: boolean;
+    presentationType: 'text' | 'date';
+    suggestionBoxAlgn: SuggestionBoxAlign | undefined;
+}
+export const defaultState: DefaultState = {
+    focused: false,
+    cleared: true,
+    matchedPretext: [],
+    items: [],
+    terms: [],
+    components: [],
+    selection: '',
+    selectionIndex: 0,
+    allowDividers: true,
+    presentationType: 'text',
+    suggestionBoxAlgn: undefined,
+};
+
+export type SuggestionBoxForwarded = {
+    getTextbox: () => HTMLInputElement | HTMLTextAreaElement | null;
+    focus: () => void;
+    blur: () => void;
+}
+
+const SuggestionBoxComponent: React.ForwardRefRenderFunction<SuggestionBoxForwarded, SuggestionBoxProps> = ({
+    listPosition = 'top',
+    containerClass = '',
+    renderDividers = [],
+    renderNoResults = false,
+    shouldSearchCompleteText = false,
+    completeOnTab = true,
+    isRHS = false,
+    requiredCharacters = 1,
+    openOnFocus = false,
+    openWhenEmpty = false,
+    replaceAllInputOnSelect = false,
+    listenForMentionKeyClick = false,
+    forceSuggestionsWhenBlur = false,
+    value,
+    onChange,
+    onItemSelected,
+    onSuggestionsReceived,
+    providers,
+    actions,
+    onKeyPress,
+    onComposition,
+    onSelect,
+    onKeyDown,
+    onFocus,
+    onBlur,
+    dateComponent,
+    listComponent,
+    contextId,
+
+    // onHeightChange,
+    ...props
+}: SuggestionBoxProps, ref) => {
+    const suggestionReadOut = React.useRef<HTMLDivElement>(null);
+    const inputRef = React.useRef<HTMLInputElement | HTMLTextAreaElement>(null);
+    const containerRef = React.useRef<HTMLDivElement>(null);
+
+    // Keep track of whether we're composing a CJK character so we can make suggestions for partial characters
+    const [isComposing, setIsComposing] = useState(false);
+    const [pretext, setPretext] = useState('');
+
+    // const pretext = '';
+
+    // Used for debouncing pretext changes
+    // let timeoutId: NodeJS.Timeout | undefined;
+    const [timeoutId, setTimeoutId] = useState<NodeJS.Timeout | undefined>(undefined);
+
+    // Used for preventing suggestion list to close when scrollbar is clicked
+    const [preventSuggestionListCloseFlag, setPreventSuggestionListCloseFlag] = useState(false);
+
+    // pretext: the text before the cursor
+    // matchedPretext: a list of the text before the cursor that will be replaced if the corresponding autocomplete term is selected
+    // terms: a list of strings which the previously typed text may be replaced by
+    // items: a list of objects backing the terms which may be used in rendering
+    // components: a list of react components that can be used to render their corresponding item
+    // selection: the term currently selected by the keyboard
+    const [state, setState] = useState(() => defaultState);
```

Comment:
I agree, they can probably be broken up. I kept them together here so it could be passed around much easier without passing multiple items into some functions. Let me break them out and see what I can do.